### PR TITLE
Add special case for integer scalars to jax.numpy.power.

### PR DESCRIPTION
**Pull request by hawkinsp**

### CODE REVIEW

**Comment by hawkinsp** [comment]
I have somewhat mixed feelings about this, given its behavior isn't invariant to `jit`, but it does seem likely to improve cases like `x ** 2` if `pow` has low precision.

@odashi

**Comment by shoyer** [comment]
NumPy uses some similar tricks, special casing exponents of 1, -1, 0, 0.5 and 2:
https://github.com/numpy/numpy/blob/3873ec46816f3fe0f705394242f83a764d4ce90e/numpy/core/src/multiarray/number.c#L468

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -524,6 +524,21 @@ def _float_divmod(x1, x2):
 
 @_wraps(onp.power)
 def power(x1, x2):
+  # Special case for small positive integer scalars: use binary exponentiation.
+  # Using lax.pow may be imprecise for floating-point values; the goal of this
+  # code path is to make sure we end up with a precise output for the common
+  # pattern ``x ** 2`` or similar.
+  if isinstance(x2, int) and x2 >= 0 and x2 <= 64:
```
Comment (shoyer) [inline jax/numpy/lax_numpy.py:531]: You _could_ safely use Python's ternary comparison here:
```suggestion
  if isinstance(x2, int) and 0 <= x2 <= 64:
```

**Comment by hawkinsp** [comment]
This fixes a test failure of `optix_test.py` so I'm planning to merge this.

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -524,6 +524,21 @@ def _float_divmod(x1, x2):
...
+    x1 = asarray(x1)
+    acc = None
+    while x2 > 0:
+      if x2 & 1:
+        acc = x1 if acc is None else lax.mul(acc, x1)
+      x2 >>= 1
+      if x2 > 0:
+        x1 = lax.mul(x1, x1)
```
Comment (shoyer) [inline jax/numpy/lax_numpy.py:539]: I would be reassured if you added a test for all integer powers in some range, e.g., 0 to 8 inclusive.

Comment (hawkinsp) [reply jax/numpy/lax_numpy.py:539]: I already did ...?

File: jax/numpy/lax_numpy.py (hunk 1 above)
Comment (hawkinsp) [reply jax/numpy/lax_numpy.py]: Done.

File: jax/numpy/lax_numpy.py (hunk 2 above)
Comment (shoyer) [reply jax/numpy/lax_numpy.py:539]: yep I was looking at the older commit

**Review by shoyer — APPROVED** [review]

**Pull request merged**